linera-core/src/chain_worker/state/attempted_changes.rs

```diff
@@ -348,26 +351,56 @@ where
         let mut actions = self.state.create_network_actions().await?;
         tracing::trace!(
             "Processed confirmed block {} on chain {:.8}",
-            block.height,
+            block_height,
             block.chain_id
         );
         actions.notifications.push(Notification {
             chain_id: block.chain_id,
             reason: Reason::NewBlock {
-                height: block.height,
+                height: block_height,
                 hash: certificate.value.hash(),
             },
         });
         // Persist chain.
         self.save().await?;
+
         self.state
             .recent_hashed_certificate_values
             .insert(Cow::Owned(certificate.value))
             .await;
 
+        self.register_delivery_notifier(block_height, &actions, notify_when_messages_are_delivered)
+            .await;
+
         Ok((info, actions))
     }
 
+    /// Schedules a notification for when cross-chain messages are delivered up to the given
+    /// `height`.
+    #[tracing::instrument(level = "trace", skip(self, notify_when_messages_are_delivered))]
+    async fn register_delivery_notifier(
```

Comment:
This is a weird method - its whole purpose is dependent on whether `notify_when_*` is `Some(_)` or `None`. I haven't read the whole upstream code but IMHO this is a code smell. We push this variable through multiple stack layers only to, possible, not even use it (if it's `None`).

Reply:
I agree it's not great to have the method depend on if the type is `Some` or not. However, pushing it through multiple stack layers is on purpose, we need to ensure that it happens in the right sequence, so it must happen in the chain worker (that's the point of this PR).

As for the method itself not being great, I'm not particularly worried because:

1. The method is private
2. It's only called from one place (the method right above it)
3. Inlining it all or just the `if let Some(_) = ...` part would make the caller function even longer and harder to read, which is especially bad because the caller function is critical to consensus